File: Day3/move_zeroes.py
"""
Move Zeroes (Easy)

Approach: Two-pointer compaction of non-zeros then fill trailing zeros.
Time: O(n), Space: O(1), minimal writes.
"""
from typing import List


def move_zeroes(nums: List[int]) -> None:
    """Move all zeros to the end in-place while preserving relative order.

    Do not return anything; modify nums in-place.
    """
    write = 0
    for read in range(len(nums)):
        if nums[read] != 0:
            nums[write] = nums[read]
            write += 1
    for i in range(write, len(nums)):
        nums[i] = 0


if __name__ == "__main__":  # simple smoke tests
    arr = [0, 1, 0, 3, 12]
    move_zeroes(arr)
    assert arr == [1, 3, 12, 0, 0]

    arr2 = [0]
    move_zeroes(arr2)
    assert arr2 == [0]

    arr3 = [1, 2, 3]
    move_zeroes(arr3)
    assert arr3 == [1, 2, 3]

    arr4 = [0, 0, 0]
    move_zeroes(arr4)
    assert arr4 == [0, 0, 0]

    arr5 = [4, 0, 5, 0, 0, -1]
    move_zeroes(arr5)
    assert arr5 == [4, 5, -1, 0, 0, 0]

    print("move_zeroes.py: all tests passed")

File: learnLeetCode/Day3/move_zeroes.cpp
#include <iostream>
#include <vector>
using namespace std;

// Move Zeroes (Easy)
// Approach: Single-pass two pointers with conditional swap for minimal writes.
// Time: O(n), Space: O(1).
void moveZeroes(vector<int>& nums) {
    int write = 0;
    for (int read = 0; read < (int)nums.size(); ++read) {
        if (nums[read] != 0) {
            if (read != write) {
                // swap keeps relative order and minimizes writes
                int tmp = nums[write];
                nums[write] = nums[read];
                nums[read] = tmp;
            }
            ++write;
        }
    }
}

int main() {
    vector<int> nums{0,1,0,3,12};
    moveZeroes(nums);
    cout << "[";
    for (size_t i = 0; i < nums.size(); ++i) {
        if (i) cout << ", ";
        cout << nums[i];
    }
    cout << "]\n";

    // basic checks
    vector<int> a{0}; moveZeroes(a); if (a != vector<int>({0})) return 1;
    vector<int> b{1,2,3}; moveZeroes(b); if (b != vector<int>({1,2,3})) return 1;
    vector<int> c{0,0,0}; moveZeroes(c); if (c != vector<int>({0,0,0})) return 1;
    vector<int> d{4,0,5,0,0,-1}; moveZeroes(d); if (d != vector<int>({4,5,-1,0,0,0})) return 1;
    return 0;
}
